### PR TITLE
Numpy array conversion in pyfunc scoring server

### PR DESCRIPTION
**Pull request by lennon310**
## What changes are proposed in this pull request?

Currently the REST POST input data is converted to Pandas Dataframe, which takes ~ 5 ms compared with ~ 1 ms on Numpy array conversion. In this PR I added the Numpy conversion from both split oriented JSON data.
  
More details: https://github.com/mlflow/mlflow/issues/1403
 
## How is this patch tested?
 
Unit tests all passed. 

The unit tests on new conversions are added.

The profiler was used locally, and it showed it takes less than 1 ms in Numpy conversion.
 
### Is this a user-facing change? 

- [ ] No. You can skip the rest of this section.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.

Please refer to https://github.com/mlflow/mlflow/issues/1403
 
### What component(s) does this PR affect?
 
- [ ] UI
- [ ] CLI 
- [ ] API 
- [ ] REST-API 
- [ ] Examples 
- [ ] Docs
- [ ] Tracking
- [ ] Projects 
- [x] Pyfunc
- [ ] Artifacts 
- [ ] Models 
- [ ] Scoring 
- [ ] Serving
- [ ] R
- [ ] Java
- [ ] Python

### How should the PR be classified in the release notes? Choose one:
 
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [x] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/pyfunc/scoring_server/__init__.py [hunk 1]
```diff
@@ -45,14 +45,18 @@
 
 CONTENT_TYPE_CSV = "text/csv"
 CONTENT_TYPE_JSON = "application/json"
-CONTENT_TYPE_JSON_RECORDS_ORIENTED = "application/json; format=pandas-records"
```
Comment (ahutterTA) [inline mlflow/pyfunc/scoring_server/__init__.py]: Why is this change necessary?

File: tests/pyfunc/test_scoring_server.py [hunk 2]
```diff
@@ -153,6 +153,32 @@ def test_scoring_server_successfully_evaluates_correct_dataframes_with_pandas_sp
     assert response.status_code == 200
 
 
+@pytest.mark.large
+def test_scoring_server_successfully_evaluates_correct_records_to_numpy(
```
Comment (ahutterTA) [inline tests/pyfunc/test_scoring_server.py]: nit: test_scoring_server_successfully_evaluates_correct_dataframes_with_pandas_records_orientation_parsed_to_numpy_array

File: tests/pyfunc/test_scoring_server.py [hunk 3]
```diff
@@ -153,6 +153,32 @@ def test_scoring_server_successfully_evaluates_correct_dataframes_with_pandas_sp
...
+    pandas_records_content = pd.DataFrame(sklearn_model.inference_data).to_json(orient="records")
+    response_records_content_type = pyfunc_serve_and_score_model(
+            model_uri=os.path.abspath(model_path),
+            data=pandas_records_content,
+            content_type=pyfunc_scoring_server.CONTENT_TYPE_JSON_RECORDS_NUMPY)
+    assert response_records_content_type.status_code == 200
+
+
+@pytest.mark.large
+def test_scoring_server_successfully_evaluates_correct_split_to_numpy(
```
Comment (ahutterTA) [inline tests/pyfunc/test_scoring_server.py]: nit: test_scoring_server_successfully_evaluates_correct_dataframes_with_pandas_split_orientation_parsed_to_numpy_array

File: mlflow/pyfunc/scoring_server/__init__.py [hunk 4]
```diff
@@ -140,8 +184,8 @@ def ping():  # pylint: disable=unused-variable
     def transformation():  # pylint: disable=unused-variable
         """
         Do an inference on a single batch of data. In this sample server,
-        we take data as CSV or json, convert it to a Pandas DataFrame,
-        generate predictions and convert them back to CSV.
+        we take data as CSV or json, convert it to a Pandas DataFrame or Numpy,
+        generate predictions and convert them back to CSV/json.
```
Comment (ahutterTA) [inline mlflow/pyfunc/scoring_server/__init__.py]: Looking at line 227 doesn't this only return json?

File: tests/pyfunc/test_scoring_server.py [hunk 5]
```diff
@@ -190,6 +216,34 @@ def test_parse_json_input_split_oriented():
     assert all(p1 == p2)
 
 
+@pytest.mark.large
+def test_parse_json_input_records_oriented_to_numpy():
```
Comment (ahutterTA) [inline tests/pyfunc/test_scoring_server.py]: nit: _to_numpy_array?

File: tests/pyfunc/test_scoring_server.py [hunk 6]
```diff
@@ -190,6 +216,34 @@ def test_parse_json_input_split_oriented():
...
+    p0 = pd.DataFrame.from_dict(data)
+    p1 = np.array([[a, b, c] for a, b, c in zip(data['col_m'], data['col_z'], data['col_a'])],
+                  dtype=object)
+    p2 = pyfunc_scoring_server.parse_records_oriented_json_input_to_numpy(
+        p0.to_json(orient="records"))
+    np.testing.assert_array_equal(p1, p2)
+
+
+@pytest.mark.large
+def test_parse_json_input_split_oriented_to_numpy():
```
Comment (ahutterTA) [inline tests/pyfunc/test_scoring_server.py]: ditto

File: tests/pyfunc/test_scoring_server.py [hunk 7]
```diff
@@ -215,6 +269,33 @@ def test_split_oriented_json_to_df():
     assert set(str(dt) for dt in df.dtypes) == {'object', 'float64', 'int64'}
 
 
+@pytest.mark.large
+def test_records_oriented_json_to_numpy():
```
Comment (ahutterTA) [inline tests/pyfunc/test_scoring_server.py]: ditto

File: tests/pyfunc/test_scoring_server.py [hunk 8]
```diff
@@ -215,6 +269,33 @@ def test_split_oriented_json_to_df():
...
+           ']'
+    df = pyfunc_scoring_server.parse_records_oriented_json_input_to_numpy(jstr)
+    np.testing.assert_array_equal(np.array([['95120', 10.45, 8],
+                                            ['95128', 23.0, 0],
+                                            ['95128', 12.1, 10]], dtype=object),
+                                  df)
+
+
+@pytest.mark.large
+def test_split_oriented_json_to_numpy():
```
Comment (ahutterTA) [inline tests/pyfunc/test_scoring_server.py]: ditto

File: mlflow/pyfunc/scoring_server/__init__.py [hunk 9]
```diff
@@ -45,14 +45,18 @@
 
 CONTENT_TYPE_CSV = "text/csv"
 CONTENT_TYPE_JSON = "application/json"
-CONTENT_TYPE_JSON_RECORDS_ORIENTED = "application/json; format=pandas-records"
-CONTENT_TYPE_JSON_SPLIT_ORIENTED = "application/json; format=pandas-split"
+CONTENT_TYPE_JSON_RECORDS_ORIENTED = "application/json-pandas-records"
+CONTENT_TYPE_JSON_SPLIT_ORIENTED = "application/json-pandas-split"
+CONTENT_TYPE_JSON_RECORDS_NUMPY = "application/json-numpy-records"
```
Comment (ahutterTA) [inline mlflow/pyfunc/scoring_server/__init__.py]: Don't you need to update the docs for this: https://mlflow.org/docs/latest/models.html#deploy-mlflow-models

File: mlflow/pyfunc/scoring_server/__init__.py (hunk 1 above)
Comment (lennon310) [reply mlflow/pyfunc/scoring_server/__init__.py]: It did not work locally for me ---- looks like the content type was truncated by the semi-colon.  So if I send something like
`curl -X POST localhost:5000/invocations -H "Content-Type: application/json; format=pandas-records" -d"@json.txt"` it actually treated if as `"Content-Type: application/json"`

File: tests/pyfunc/test_scoring_server.py (hunk 2 above)
Comment (lennon310) [reply tests/pyfunc/test_scoring_server.py]: The name is too long to fit in one line :)

File: mlflow/pyfunc/scoring_server/__init__.py (hunk 4 above)
Comment (lennon310) [reply mlflow/pyfunc/scoring_server/__init__.py]: good catch

File: mlflow/pyfunc/scoring_server/__init__.py (hunk 9 above)
Comment (lennon310) [reply mlflow/pyfunc/scoring_server/__init__.py]: I'm not sure if I need to, I will check with Databricks team on this.

File: mlflow/pyfunc/scoring_server/__init__.py [hunk 10]
```diff
@@ -45,14 +46,18 @@
 
 CONTENT_TYPE_CSV = "text/csv"
 CONTENT_TYPE_JSON = "application/json"
-CONTENT_TYPE_JSON_RECORDS_ORIENTED = "application/json; format=pandas-records"
-CONTENT_TYPE_JSON_SPLIT_ORIENTED = "application/json; format=pandas-split"
+CONTENT_TYPE_JSON_RECORDS_ORIENTED = "application/json-pandas-records"
```
Comment (aarondav) [inline mlflow/pyfunc/scoring_server/__init__.py]: Is this a backwards-incompatible change? We probably shouldn't break old clients.

Comment (lennon310) [reply mlflow/pyfunc/scoring_server/__init__.py]: Is there a way we can test? I can revert the two lines of pandas related changes, however, I don't think it works according to my local test. Please take a look here:

https://github.com/mlflow/mlflow/pull/1479#discussion_r295825703

Comment (lennon310) [reply mlflow/pyfunc/scoring_server/__init__.py]: Talked offline with @aarondav , he couldn't reproduce my error. Reverting to
```
CONTENT_TYPE_JSON_RECORDS_ORIENTED = "application/json; format=pandas-records"
CONTENT_TYPE_JSON_SPLIT_ORIENTED = "application/json; format=pandas-split"
```

**Comment by tomasatdatabricks** [comment]
Hi @lennon310, thanks for the contribution!

In general we would like to keep the model interface consistent. It would be great if we can find a way to improve the performance but preserve the existing interface. I noticed the pandas read_json has a numpy flag which set to false by default. If set, it will decode all data directly to numpy arrays. I think we should consider this instead. If it's not too much slower then setting the numpy flag would be preferable to custom parsing. If it is slower, can we wrap the numpy array in a pandas DataFrame? That should be zero copy and low overhead.

Can you run your benchmark with the two options ^^ and update the PR?

On a separate note, can you describe what is your motivation for this use case? We are considering adding support for binary data format for faster performance. Would that be interesting to you?

**Comment by lennon310** [comment]
Thanks @tomasatdatabricks for your suggestion, I understand your concern. 

However, setting `numpy=True` seems not help. I updated the profiler result in the [original issue](https://github.com/mlflow/mlflow/issues/1403)

**Comment by tomasatdatabricks** [comment]
Hi @lennon310, thanks for the reply and for updating the benchmark. 

I actually looked at the parsing code in pandas and it appears to hardcode the numpy flag to false when the orient == split, so no wonder this did not help. Can you try with orient = "records"?

The binary data should help with both data size and conversion cost since they typically do not need to be parsed. 

As for the wrapping of  numpy as DataFrame, can you try something like this:

```
pandas.DataFrame(index=json_input_list["index"], data=numpy.array(json_input_list['data']), columns=json_input_list["columns"])
```

**Comment by lennon310** [comment]
@tomasatdatabricks  Thank you very much for your suggestion!
I profiled locally and found both wrapping numpy to pandas and setting `numpy=True` reduced the conversion time (~ 2 ms, updated in issue).  However, it says numpy option supports numeric data only, so it will break the unit tests.

As a result, I'm applying your 2nd suggestion (wrapping numpy into pandas). Please take another look. Thanks

**Comment by tomasatdatabricks** [comment]
Thanks for the updates @lennon310!

I have two more questions / requests.

1) Why do you set dtype=object? Is that required for your use case? I would prefer if we did not have to specify that because I don't think this would work with many ml frameworks (e.g. tensorflow would probably want numbers as numbers).  

If you do need to specify it, can we add it to the header specifying input type and leave it empty by default? 

2) The case for orient=records should also specify column names. This is tricky because we assume the data comes in consistent order while json objects are unordered. Can we verify the order somehow without compromising performance ? E.g. something like (+ test verifying that we do detect inconsistent ordering):

```
columns = json_input_list[0].keys()
for d in json_input_list:
  if d.keys() != columns:
    raise Exception("Inconsistent order in the json input.")
 return pd.DataFrame(columns=columns, data=np.array([list(d.values()) for d in json_input_list]))
```

Alternatively, I would consider enabling the numpy parsing only for orient=split. I think that would be better than having silent failures.

**Comment by lennon310** [comment]
Thanks @tomasatdatabricks .
1. That dtype is set during numpy conversion but not pandas dataframe creation, so I think it should be fine maybe?  Without setting `dtype=object` the numpy may treat the elements as string (i.e., in `test_split_oriented_json_to_numpy_array` the array will be `array([['95120', '10.45', '-8'],
              ['95128', '23.0', '-1'],
              ['95128', '12.1', '1000']]` instead of `array([['95120', 10.45, -8],
              ['95128', 23.0, -1],
              ['95128', 12.1, 1000]]`).

2. I was setting `object_pairs_hook=OrderedDict` and I thought it should be able to persist the order? I could add the order check but it might increase the latency again.

**Comment by lennon310** [comment]
@tomasatdatabricks `columns` is specified in records oriented Json case

**Comment by lennon310** [comment]
@tomasatdatabricks I was re-thinking of the dtype config, maybe we could remove it.  The dtype conversion, if necessary, could be implemented in the overriding of `predict` function.  PR updated.

**Comment by tomasatdatabricks** [comment]
Thanks for the quick updates @lennon310!

1) I want to look into what would be the implications of setting dtype=object. Not specifying dtype works nicely for datasets with uniform data type, however, if you have a mix of string and number columns (which is very common actually) then the data would be parsed as strings. I am not sure what is the right thing to do (I think the ideal solution would be to pass data as columns instead of rows but that's out of scope of this issue).

2) As for the order of columns for orient=records - the dictionary just preserves the ordering of the given row, but it does not enforce uniformity in the whole dataset. e.g.

```
'[{"a":1, b:2}, {"b":3, "a":2}]"  
```

would parse without errors as:

```
[[1, 2], [3, 2]]
```

I think we should check that the column ordering matches for all rows, or just not support this feature for records format.

**Comment by lennon310** [comment]
@tomasatdatabricks I'm updating the PR and resolving your 2nd comment for now.
Basically that could be some issue from the user by messing up the keys order. And it is causing problem ---- I have to use ordered dict to keep rows order, but `odict_keys(['a', 'b']) == odict_keys(['b', 'a'])` will return `True` so there is not a good way to make sure the keys are in order. I will remove the records oriented json conversion for now.  Thanks!

**Review by tomasatdatabricks — COMMENTED** [review]
Thanks for the contribution and all the updates! lgtm
